File: src/native/RoomPlanScanner.ts
import { Platform } from 'react-native';

import ExpoRoomPlanScanner, {
  RoomPlanErrorCode,
  RoomScanResult,
  RoomScanSummary,
} from '../../modules/expo-room-plan-scanner/src';

export type { RoomScanResult, RoomScanSummary, RoomPlanErrorCode };

export class RoomPlanError extends Error {
  readonly code: RoomPlanErrorCode | string;

  constructor(code: RoomPlanErrorCode | string, message: string) {
    super(message);
    this.name = 'RoomPlanError';
    this.code = code;
  }
}

/**
 * Resolves true only on iOS 16+ devices with a LiDAR scanner.
 * Always false on Android, simulator, and unsupported iPhones/iPads.
 */
export async function isRoomPlanSupported(): Promise<boolean> {
  if (Platform.OS !== 'ios') return false;
  try {
    return await ExpoRoomPlanScanner.isRoomPlanSupported();
  } catch (err) {
    console.warn('[RoomPlanScanner] isRoomPlanSupported threw:', err);
    return false;
  }
}

/**
 * Presents the native RoomPlan scanner UI modally. Resolves with the captured
 * geometry when the user finishes, or rejects with `RoomPlanError` if the user
 * cancels, the device is unsupported, or processing fails.
 */
export async function startRoomScan(roomId: string): Promise<RoomScanResult> {
  if (Platform.OS !== 'ios') {
    throw new RoomPlanError('E_UNSUPPORTED_OS', 'RoomPlan is only available on iOS.');
  }
  if (!roomId) {
    throw new RoomPlanError('E_INVALID_ARGUMENT', 'A roomId is required to start a scan.');
  }
  try {
    return await ExpoRoomPlanScanner.startRoomScan(roomId);
  } catch (err) {
    const e = err as { code?: string; message?: string };
    throw new RoomPlanError(e.code ?? 'E_SCAN_FAILED', e.message ?? 'Scan failed.');
  }
}

export default {
  isRoomPlanSupported,
  startRoomScan,
};

File: modules/expo-room-plan-scanner/ios/RoomPlanScanner.swift
import ExpoModulesCore
import UIKit

#if canImport(RoomPlan)
import RoomPlan
#endif

/// Expo Module bridge that exposes Apple's RoomPlan scanner to React Native.
///
/// JS-facing API (see `modules/expo-room-plan-scanner/src/types.ts`):
///   - isRoomPlanSupported(): Promise<boolean>
///   - startRoomScan(roomId: string): Promise<RoomScanResult>
///
/// All scanner functionality requires iOS 16.0+ and a LiDAR-equipped device.
/// The module is built with availability + canImport guards so it compiles
/// on older SDKs and gracefully reports `E_UNSUPPORTED_OS` /
/// `E_UNSUPPORTED_DEVICE` at runtime.
public class RoomPlanScanner: Module {
  public func definition() -> ModuleDefinition {
    // Name JS uses: requireNativeModule('ExpoRoomPlanScanner')
    Name("ExpoRoomPlanScanner")

    AsyncFunction("isRoomPlanSupported") { () -> Bool in
      if #available(iOS 16.0, *) {
        #if canImport(RoomPlan)
        return RoomCaptureSession.isSupported
        #else
        return false
        #endif
      }
      return false
    }

    AsyncFunction("startRoomScan") { (roomId: String, promise: Promise) in
      DispatchQueue.main.async {
        guard #available(iOS 16.0, *) else {
          promise.reject("E_UNSUPPORTED_OS", "RoomPlan requires iOS 16.0 or newer.")
          return
        }

        #if canImport(RoomPlan)
        guard RoomCaptureSession.isSupported else {
          promise.reject(
            "E_UNSUPPORTED_DEVICE",
            "This device does not support RoomPlan. A LiDAR-equipped iPhone or iPad is required."
          )
          return
        }

        guard let presenter = Self.topMostViewController() else {
          promise.reject("E_NO_PRESENTER", "Could not find a view controller to present the scanner.")
          return
        }

        let scannerVC = RoomCaptureViewController(roomId: roomId) { result in
          presenter.dismiss(animated: true) {
            switch result {
            case .success(let payload):
              promise.resolve(payload)
            case .failure(let error):
              let nsError = error as NSError
              let code = nsError.userInfo["code"] as? String ?? "E_SCAN_FAILED"
              promise.reject(code, nsError.localizedDescription)
            }
          }
        }
        scannerVC.modalPresentationStyle = .fullScreen
        presenter.present(scannerVC, animated: true)
        #else
        promise.reject("E_FRAMEWORK_MISSING", "RoomPlan framework is not available in this build.")
        #endif
      }
    }
  }

  /// Walks the active scene to find the top-most view controller suitable for presentation.
  /// Uses the modern UIScene API (avoids deprecated `UIApplication.shared.windows`).
  private static func topMostViewController() -> UIViewController? {
    let scenes = UIApplication.shared.connectedScenes
      .compactMap { $0 as? UIWindowScene }
      .filter { $0.activationState == .foregroundActive }

    let keyWindow = scenes
      .flatMap(\.windows)
      .first(where: \.isKeyWindow) ?? scenes.first?.windows.first

    var topVC = keyWindow?.rootViewController
    while let presented = topVC?.presentedViewController {
      topVC = presented
    }
    return topVC
  }
}
